bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/AbstractZkLedgerManagerFactory.java

```diff
@@ -194,15 +200,78 @@ public static LedgerManagerFactory newLedgerManagerFactory(
                 }
                 factoryClass = theCls.asSubclass(LedgerManagerFactory.class);
             } catch (ClassNotFoundException cnfe) {
-                throw new IOException("Failed to instantiate ledger manager factory "
-                        + layout.getManagerFactoryClass());
+                factoryClass = attemptToResolveShadedLedgerManagerFactory(
+                    conf,
+                    layout.getManagerFactoryClass(),
+                    cnfe);
             }
         }
         // instantiate a factory
         lmFactory = ReflectionUtils.newInstance(factoryClass);
         return lmFactory.initialize(conf, layoutManager, layout.getManagerVersion());
     }
 
+    private static String normalizedLedgerManagerFactoryClassName(String factoryClass,
+                                                                  String shadedClassPrefix,
+                                                                  boolean isShadedClassAllowed) {
+        if (isShadedClassAllowed) {
+            if (null == factoryClass || null == shadedClassPrefix) {
+                return factoryClass;
+            } else {
+                return factoryClass.replace(shadedClassPrefix, "");
+            }
+        } else {
+            return factoryClass;
+        }
+    }
+
+    private static boolean isSameLedgerManagerFactory(
+            AbstractConfiguration<?> conf, String leftFactoryClass, String rightFactoryClass) {
+        leftFactoryClass = normalizedLedgerManagerFactoryClassName(
```

Comment:
```
return leftFactoryClass.endsWith(rightFactoryClass) || rightFactoryClass.endsWith(leftFactoryClass)
```

Reply:
you can't do simply do that, because left or right can be null and it needs to take the flag into account.

Reply:
sure.